### ISSUE TITLE
grpc 1.0.1 shared compile fails on MinGW64 / MSYS2

### ISSUE THREAD

**Issue by emmenlau** [grpc/grpc#6072]
It seems that on MinGW, one needs a def-file for each library to be linked. As an example, see the Makefile section for gpr:

```
ifeq ($(SYSTEM),MINGW32)
$(LIBDIR)/$(CONFIG)/gpr$(SHARED_VERSION).$(SHARED_EXT): $(LIBGPR_OBJS)  $(ZLIB_DEP)
    $(E) "[LD]      Linking $@"
    $(Q) mkdir -p `dirname $@`
    $(Q) $(LD) $(LDFLAGS) -L$(LIBDIR)/$(CONFIG) -shared gpr.def -Wl,--output-def=$(LIBDIR)/$(CONFIG)/gpr$(SHARED_VERSION).def -Wl,--out-implib=$(LIBDIR)/$(CONFIG)/libgpr$(SHARED_VERSION)-dll.a -o $(LIBDIR)/$(CONFIG)/gpr$(SHARED_VERSION).$(SHARED_EXT) $(LIBGPR_OBJS) $(LDLIBS) $(ZLIB_MERGE_LIBS)
else
```

gpr.dll requires gpr.def, just as grpc requires grpc.def. However it seems that the shared library def's exist only for grpc, not for gpr, grpc_unsecure, grpc_zookeeper, grpc++, grpc++_unsecure, and grpc_csharp_ext. Is this by mistake, or is the shared library building not supported for these targets?

**Comment by LukeMauldin** [comment]
I think I maybe experiencing something similar: #6058

**Comment by emmenlau** [comment]
@LukeMauldin  this could be related, but as far as I recall the Makefile explicitly specifies the gpr.def for MinGW32, opposed to other architectures. I might be wrong though.

**Comment by soltanmm-google** [comment]
Woosh. Up goes the version number.

**Comment by emmenlau** [comment]
Wait, is the issue fixed? Or just closed?

**Comment by hsaliak** [comment]
@emmenlau the issue just closed. Are you still experiencing this with v.1.0.0 release?

**Comment by hsaliak** [comment]
@emmenlau it was just closed -- since this is not a platform we are supporting actively, however, we're happy to get some help in verifying if this is still an issue with HEAD and/or 1.0.x and patches that can address the issue, would you be able to help?

**Comment by emmenlau** [comment]
Issue still exists with grpc 1.0.1:
```
[LD]      Linking /d/Temp/mingw-w64-grpc/src/build-x86_64-w64-mingw32/libs/opt/gpr-1.dll
cc.exe: error: gpr.def: No such file or directory
```
I think this would be easy to solve: you already have the def-file for grpc: grpc.def

Please add similarly def-files for gpr, grpc_unsecure, grpc_zookeeper, grpc++, grpc++_unsecure, and grpc_csharp_ext.

**Comment by hsaliak** [comment]
Would you be able to send a pull request? :)

**Comment by emmenlau** [comment]
Dear @hsaliak, I spend now several hours making grpc 1.0.1 compile on MSYS2 with gcc-6.2.0. It would be great if you can integrate my findings into your build system. Attached is a patch that makes the compile work. The patch includes multiple changes that I describe here:

 - The missing def-files might *not* be required. However, they are referenced in your Makefile. This breaks the build. So in the Makefile, where the linker command for shared libraries reads '-shared <libname>.def' you need to remove the '<libname>.def'.
 - The options used for strip are too aggressive. In the Makefile in line 8, replace 'STRIP ?= strip' with 'STRIP ?= strip --strip-unneeded'. This is also reported here: https://github.com/grpc/grpc/issues/6136
 - The Makefile sets variable 'SHARED_VERSION = -1' for MINGW32. This is not required and not helpful. Better set this variable to empty like on other platforms: 'SHARED_VERSION ='
 - The Makefile sets many static library names with a suffix '-imp', like 'libgrpc++_unsecure-imp.a'. This is not required and not helpful. Better remove the '-imp' suffix.
 - There is the issue I've reported here: https://github.com/grpc/grpc/issues/8774
 - Finally the issue I've reported here: https://github.com/grpc/grpc/issues/6073

Attached patch includes all these changes. However, since your Makefile is auto-generated, you might need to abstract the changes into your build system.
[0001-grpc-mingw64-split.patch.txt](https://github.com/grpc/grpc/files/597085/0001-grpc-mingw64-split.patch.txt)

**Comment by emmenlau** [comment]
Dear @hsaliak and all, do you think this can be integrated into grpc? I've been plagued by these issues since at least grpc-0.12.0. With my proposed changes, MSYS2 (and most likely also MinGW64) would work fine!

**Comment by jtattermusch** [comment]
@emmenlau, can you please try to come up with a pull request that includes the changes you're requesting? Updating the https://github.com/grpc/grpc/blob/master/templates/Makefile.template shouldn't be very hard (run tools/buildgen/generate_projects.sh) to regenerate. Then we can assign you a reviewer (probably @nicolasnoble) and we can discuss the actual changes that are being done.

Also, please not that this week we are in C core fixit mode (changes not fixing existing errors reported by our CI system are being postponed until next week) - but that gives you some time to come up with that PR.

Thanks!

**Comment by emmenlau** [comment]
Dear @jtattermusch and @hsaliak , I've implemented a number of my proposed changes in four pull requests. But I have not implemented all my proposed changes yet. It would be super awesome if you can review/comment/validate my PR's before I continue, so I know I'm on the right track.
I plan to implement full support for MSYS2, and its mostly minor changes in the Makefile.template, plus a few minor compiler-related code cleanups (removing unused variables etc).

**Comment by hsaliak** [comment]
@nicolasnoble would you be able to help review as well?

**Comment by hsaliak** [comment]
@emmenlau  Just want to confirm that thePRs that you want reviewed are:
#9072 #9073 #9075 #9076 and #9096 in that order..

**Comment by emmenlau** [comment]
@hsaliak yes thanks for pointing them out! So its five, not four :-) Feedback would be greatly appreciated.

**Comment by jtattermusch** [comment]
@emmenlau, sorry for the delay, I took an initial look at your PRs, they mostly look good, thanks for the contributions. For some of the makefile changes, I requested @nicolasnoble 's opinion.

**Comment by emmenlau** [comment]
Dear @jtattermusch and all, thanks for getting this on the way! I have a few more (equally minor) changes in the pipeline. Afterwards I hope grpc should build cleanly on MSYS2.

**Comment by jtattermusch** [comment]
Fixed in #9307.
